Proj1 multiple asic cfg

J0nPatrik opened the pull request:
Prepared for two new OP codes.
One that stores ASIC configurations in NVM memory.
One that gets a stored ASIC configuration in NVM memory
and applies it to the ASIC.

patrikj5 committed:
Store ASIC configuration on NVM

Changing memory managemen so that MSP_OP_SEND_CUBES_CITI_CONF
by default stores the config on NVM memory instead of config mem.
The command does not write to config mem by default only stores in NVM.

Prepared for new command in order to apply a specific ASIC config
stored in NVM mem.

J0nPatrik committed:
Added two new OP codes

MSP_OP_SEND_NVM_CITI_CONF:
  OP command to send ASIC configuration
  to NMV memory.
MSP_OP_APPLY_CUBES_CITI_CONF:
  Send configuration number that is to
  be read in the NVM memory and applied
  to the configuration memory and AISC.

J0nPatrik committed:
Added two new OP codes

MSP_OP_SEND_NVM_CITI_CONF:
  OP command to send ASIC configuration
  to NMV memory.
MSP_OP_APPLY_CUBES_CITI_CONF:
  Send configuration number that is to
  be read in the NVM memory and applied
  to the configuration memory and AISC.

J0nPatrik committed:
Merge branch 'proj1_multiple_asic_cfg' of github.com:tstana/cubes-fw into proj1_multiple_asic_cfg

J0nPatrik committed:
Restored a OP code

Restored OP code
  MSP_OP_SEND_CUBES_PROB_CONF
that was accidentally removed.

tstana commented on mem_mgmt/mem_mgmt.h:
```diff
@@ -46,7 +46,7 @@
 #define RAM_ADDR        (0x20000000u)
 #define NVM_ADDR        (0x60000000u)
-#define CITIROC_OFFSET  (0xF010u)
+#define CITIROC_OFFSET  (0x40000u)
```
This will lead to a run-time error, as this address does not exist (at least if I check the MSS User Guide). The device present on the CUBES PCB is an M2S010 device, which does not have eNVM_1 (which would start at 0x60040000).

Suggest you use another offset.

tstana commented on msp/msp_exp_handler.c:
```diff
@@ -621,11 +621,21 @@ void msp_exprecv_complete(unsigned char opcode)
 			break;
 
+		case MSP_OP_APPLY_CUBES_CITI_CONF:
+			uint8_t *nvm_cfg_addr=(uint8_t *)(NVM_ADDR+CITIROC_OFFSET+(recv_data[0]*CITIROC_LEN));
```
This line causes a compilation error for me, because you are trying to declare a variable within a case statement. You can either add braces around the block of the case statement, or declare the `nvm_cfg_addr` pointer outside the switch statement and just assign the pointer here.

Also, please keep this line within 80 character columns.

tstana reviewed (changes requested):
Good job on the changes @J0nPatrik, it looks to me like they should work once the two things I commented on have been addressed.

patrikj5 committed:
Change offset address for CITIROC

Added check to see if requested ASIC configuration stored
in NVM is a valid configuration

patrikj5 committed:
Revert "Change offset address for CITIROC"

This reverts commit fe5b2729ed09a95777f31640021afc0e98374b9b.

J0nPatrik committed:
Changed CITIROC_OFFSET

Added check to see if the requested ASIC
configuration in NVM is a valid configuration

tstana commented on msp/msp_exp_handler.c:
```diff
@@ -626,11 +626,17 @@ void msp_exprecv_complete(unsigned char opcode)
 			break;
 
 		case MSP_OP_APPLY_CUBES_CITI_CONF:
```
I guess you forgot to change the name of the macro, should it be `MSP_OP_SELECT_NVM_CITI_CONF`?

tstana commented on mem_mgmt/mem_mgmt.h:
```diff
@@ -46,8 +46,8 @@
 #define NVM_ADDR        (0x60000000u)
-#define CITIROC_OFFSET  (0x40000u)
 #define SEQFLAG_OFFSET  (0xF100)
+#define CITIROC_OFFSET  (0xFE10)
```
I'd put this at a "rounder" address, e.g., `0xFF00`, or better yet `0x10000`?

J0nPatrik committed:
Changed CITIROC_OFFSET address
Fixed naming:
MSP_OP_APPLY_NVM_CONF -> MSP_OP_SELECT_NVM_CONF

J0nPatrik committed:
Save ASIC config number
Save the ASIC config number to address 0xFF10 in NVM when commande MSP_OP_SELECT_NVM_CONF is invoked

Joarfo committed:
proj4_startup

J0nPatrik committed:
Namechange on define saving latest ASIC configuration used

NVM_CITIROC_CONF_OFFSET -> NVM_CITIROC_CONF_NO_OFFSET

tstana commented on main.c:
```diff
@@ -69,6 +69,22 @@ int main(void)
     MSS_TIM64_enable_irq();
     NVIC_SetPriority(Timer1_IRQn, 1);
 
+    /* load citi conf on startup */
```
Why not spell this out, "Load Citiroc configuration [...]"?

tstana commented on main.c:
```diff
@@ -69,6 +69,22 @@ int main(void)
     NVIC_SetPriority(Timer1_IRQn, 1);
 
+    /* load citi conf on startup */
+    uint32_t cfg_id_32 = 0;
```
Just wondering: Why "_32"? (@anwestm, I guess this question is more directed to you.)

anwestm replied:
mem_read takes a uint32_t pointer

tstana replied:
I see. It should also work if you do `mem_read(NVM_CITIROC_CONF_NUM, (uint32_t *)&cfg_id);` on the next line. Might as well do that and test it -- if you intend to do some testing during the next couple of weeks.

tstana replied:
Fixed as part of commit 60f58db and tested to work.

tstana commented on main.c:
```diff
@@ -69,6 +69,22 @@ int main(void)
+    mem_read(NVM_CITIROC_CONF_NO, &cfg_id_32);
+    uint8_t cfg_id = (uint8_t)(cfg_id_32);
+    uint8_t *nvm_cfg_addr =
+    (uint8_t *)(NVM_ADDR+CITIROC_OFFSET+(cfg_id*CITIROC_LEN));
```
This line can be indented by a TAB or two. Alternatively, you could leave the `nvm_cfg_addr` pointer declaration without an initialization and assign to the pointer on the next line (this might look cleaner).

tstana commented on main.c:
```diff
@@ -69,6 +69,22 @@ int main(void)
+        mem_ram_write(RAM_CITI_CONF, nvm_cfg_addr);
+        citiroc_send_slow_control();
+    }
+
```
What happens in case the config is not set here? Please add a `TODO` -- and it would be greatly appreciated if you could offer some thoughts in your report. (@anwestm I guess this is also to you.)

anwestm replied:
It checks the id with the id in the configuration. if they don't match, the config will be applied. This should probably be handled anyways.

tstana replied:
Thanks for adding the TODO. It probably works as it is now in that no config will be applied to the ASIC, we should figure out in the future a means of informing "the ground" that no ASIC was applied.

As stated above, some thoughts (if you have any) on how this could be done would be greatly appreciated in your report!

tstana commented on main.c:
```diff
@@ -69,6 +69,22 @@ int main(void)
+        citiroc_send_slow_control();
+    }
+
+
+    //mem_ram_write(RAM_CITI_CONF, nvm_cfg_addr);
+    //citiroc_send_slow_control();
+
```
These lines can be removed.

tstana commented on main.c:
```diff
@@ -69,6 +69,22 @@ int main(void)
+
+    if (cfg_id == nvm_cfg_addr[CITIROC_LEN-1]) {
+        mem_ram_write(RAM_CITI_CONF, nvm_cfg_addr);
+        citiroc_send_slow_control();
```
I'm getting a compiler warning for an implicit declaration of this function. Please add an `#include` of the `citiroc.h` file.

tstana commented on mem_mgmt/mem_mgmt.c:
```diff
@@ -102,6 +106,10 @@ uint32_t mem_read(uint32_t modul, uint32_t *data)
 			break;
+		case NVM_CITIROC_CONF_NO:
+            length=CITIROC_NO_LEN;
+            addr=(uint32_t *)(NVM_ADDR+CITIROC_CONF_OFFSET);
```
This will probably not work, as `CITIROC_CONF_OFFSET` is not defined. (Check compiler warnings -- or errors -- much? :] ) Did you maybe mean `CITIROC_CONF_NO_OFFSET`?

tstana commented on mem_mgmt/mem_mgmt.h:
```diff
@@ -30,24 +30,27 @@
+#define NVM_HVPS        	(0x02u)
+#define NVM_SEQFLAG     	(0x03u)
+#define NVM_RESET       	(0x04u)
+#define NVM_CITIROC_CONF_NO	(0x05u)
```
Suggest you replace `NO` for `NUM`.

tstana commented on mem_mgmt/mem_mgmt.h:
```diff
@@ -30,24 +30,27 @@
 
 #define CITIROC_OFS     (0x00)
 #define CITIROC_LEN     (144u)
+#define CITIROC_NO_LEN	(4u)
```
Suggest you replace `NO` for `NUM`.

tstana commented on msp/msp_exp_handler.c:
```diff
@@ -621,11 +621,28 @@ void msp_exprecv_complete(unsigned char opcode)
 
+		case MSP_OP_SELECT_NVM_CITI_CONF:
+		{
+			uint8_t cfg_id = (uint8_t)recv_data[0];
```
This should be the global variable defined in `main.c`; see my comment on the `cfg_id` variable in `main.c`.

tstana commented on main.c:
```diff
@@ -69,6 +69,22 @@ int main(void)
+    /* load citi conf on startup */
+    uint32_t cfg_id_32 = 0;
+    mem_read(NVM_CITIROC_CONF_NO, &cfg_id_32);
+    uint8_t cfg_id = (uint8_t)(cfg_id_32);
```
This should be a global variable, i.e., declared outside the function and redeclared as `extern` in every file it is in use (e.g., `msp_exp_handler.c`).

tstana commented on msp/msp_opcodes.h:
```diff
@@ -43,6 +43,8 @@
 #define MSP_OP_SEND_CUBES_HVPS_TMP_VOLT			0x75
 #define MSP_OP_SEND_READ_REG_DEBUG				0x76
 #define MSP_OP_SEND_CUBES_GATEWARE_CONF			0x77
+#define MSP_OP_SEND_NVM_CITI_CONF               0x78
+#define MSP_OP_SELECT_NVM_CITI_CONF             0x79
```
These need to be redefined as was done on `proto-cubes-obc`. (The Project 2 PR which was already merged already defined `0x78`.)

tstana reviewed (changes requested):
@J0nPatrik, @Joarfo, @anwestm thanks for your work, we're getting there, but there are some changes to be made, as you can see in the comments.

Also, the `cfg_id` global variable needs to make its way to the histogram header, as we previously discussed @J0nPatrik and @Joarfo. This can be done by writing the `cfg_id` variable to `send_data_payload` at the end of `prep_data_payload()`, similar to how the `bin_cfg` is written. You can use `send_data_payload[249]` for the `cfg_id`.

anwestm committed:
- fixed tidyness of code

J0nPatrik committed:
Define namechange set cfg_id as global

Define namechange to match proto-cubes

Made variable cfg_id that stores the ASIC configuration id
global in orde to be reachable from main.c and msp_exp_handler.c

J0nPatrik committed:
Merge branch 'proj1_multiple_asic_cfg' of github.com:tstana/cubes-fw into proj1_multiple_asic_cfg

J0nPatrik committed:
Fixed name bug in main.c

J0nPatrik committed:
Added include in main.c

J0nPatrik committed:
Added full path for citiroc.h in main.c

Commit:
Remove unnecessary variable declaration

The `cfg_id_32` variable was unnecessary, it could be replaced by a
direct write to the `uint8_t cfg_id` already defined, with a cast
applied in the call to `mem_read()`.
